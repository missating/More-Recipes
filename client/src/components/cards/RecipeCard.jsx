import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';


const RecipeCard = props => (
  <div className="flex">

    <div className="card">
      <Link to={`/recipes/view/${props.id}`} className="hover">
        <img
          src={props.recipeImage}
          alt={props.recipeImage}
          className="image-width"
        />
      </Link>
      <div className="container p-3 recipedetails-column">
        <h2>{props.name}</h2>
        <p><small>{props.description}</small></p>
        <div className="center-button">
          <Link
            className="btn btn-secondary"
            to={`/recipes/view/${props.id}`}
          >
            VIEW
          </Link>
          {
            props.button &&
            (
              <button
                id="removeButton"
                className="btn btn-secondary"
                onClick={() => props.onButtonClick(props.id)}
              >
                Remove
              </button>
            )

          }
        </div>
      </div>
    </div>
  </div>
);


RecipeCard.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  recipeImage: PropTypes.string.isRequired,
  onButtonClick: PropTypes.func,
  button: PropTypes.bool
};

RecipeCard.defaultProps = {
  onButtonClick: () => null,
  button: false
};

export default RecipeCard;
