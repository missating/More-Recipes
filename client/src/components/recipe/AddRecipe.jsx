import React from 'react';
import axios from 'axios';
import Dropzone from 'react-dropzone';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';

// validations
import recipeValidator from '../../validation/recipeValidator';

// action
import addRecipe from '../../actions/addRecipe';


/**
 * @description Creates AddRecipe component
 *
 * @class AddRecipe
 *
 * @extends {React.Component}
 */
export class AddRecipe extends React.Component {
  /**
   * @description Creates an instance of AddRecipe.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      newRecipe: {
        name: '',
        ingredients: '',
        description: '',
        recipeImage: '',
        submitting: false
      },
      errors: {},
      // eslint-disable-next-line
      defaultImg: 'https://res.cloudinary.com/dxayftnxb/image/upload/v1517914951/noImage_u3sry1.png',
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
  }

  /**
   * @description Bind the value of the inputs to state
   *
   * @method onChange
   *
   * @param {object} event
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  onChange(event) {
    this.setState({
      newRecipe: {
        ...this.state.newRecipe,
        [event.target.name]: event.target.value
      }
    });
  }

  /**
   * @description Submit the form
   *
   * @method onSubmit
   *
   * @param {object} event
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  onSubmit(event) {
    event.preventDefault();
    this.setState({
      submitting: true
    });
    const { newRecipe } = this.state;
    const isValid = this.isValid();
    if (isValid) {
      if (newRecipe.recipeImage) {
        this.uploadToCloudinary().then((response) => {
          const secureUrl = response.data.secure_url;
          const recipeData = this.state.newRecipe;
          recipeData.recipeImage = secureUrl;

          this.props.addNewRecipe(recipeData)
            .then(() => {
              this.setState({
                newRecipe: {
                  name: '',
                  ingredients: '',
                  description: '',
                  recipeImage: '',
                  submitting: false,
                }
              });
              this.props.history.push('/users/recipes');
            });
        });
      } else {
        const recipeData = { ...newRecipe, recipeImage: this.state.defaultImg };
        this.props.addNewRecipe(recipeData)
          .then(() => {
            this.setState({
              newRecipe: {
                name: '',
                ingredients: '',
                description: '',
                recipeImage: '',
                submitting: false,
              }
            });
            this.props.history.push('/users/recipes');
          });
      }
    }
  }

  /**
   * @description Handles image drop
   *
   * @method handleDrop
   *
   * @memberof AddRecipe
   *
   * @param {array} files
   *
   * @returns {void}
   */
  handleDrop(files) {
    this.setState({
      newRecipe: {
        recipeImage: files[0]
      }
    });
  }


  /**
   * @description Handle image upload
   *
   * @method  uploadToCloudinary
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  uploadToCloudinary() {
    const formData = new FormData();
    formData.append('file', this.state.newRecipe.recipeImage);
    formData.append('upload_preset', 'moreRecipes');
    formData.append('api_key', '462882972372719');

    return axios.post(
      'https://api.cloudinary.com/v1_1/dxayftnxb/image/upload',
      formData, {
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
      }
    );
  }

  /**
   * @description Validates user's data before making post request
   *
   * @method isValid
   *
   * @memberof AddRecipe
   *
   * @returns {boolean} true or false
   */
  isValid() {
    const { errors, isValid } = recipeValidator(this.state.newRecipe);
    if (!isValid) {
      this.setState({ errors, submitting: false });
    } else {
      this.setState({ errors: {} });
      return isValid;
    }
  }

  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   *
   */
  render() {
    const { errors, newRecipe, defaultImg } = this.state;
    return (
      <div>
        {
          !this.props.authenticated &&
          <Redirect to="/" />
        }


        <section className="container" id="recipes">

          <h3 className="text-center mb-4">New Recipe</h3>

          <div className="row">
            <div className="col-md-12 col-sm-12">
              <form onSubmit={this.onSubmit}>
                <Dropzone
                  onDrop={this.handleDrop}
                  accept="image/*"
                  multiple={false}
                  style={{
                    border: '3px dashed ',
                    width: '300px',
                    height: 'auto',
                    margin: '0 auto',
                    padding: '5px'
                  }}
                >
                  {!newRecipe.recipeImage.preview &&
                    <div>
                      <img
                        src={defaultImg}
                        alt=""
                        className="img-fluid mx-auto d-block"
                      />
                      <h5 className="text-center">
                        Click here to upload
                      </h5>
                    </div>
                  }
                  {
                    newRecipe.recipeImage.preview &&
                    <img
                      src={newRecipe.recipeImage.preview}
                      alt=""
                      className="img-fluid mx-auto d-block"
                    />
                  }
                </Dropzone>

                <div className="form-group form-width">
                  <input
                    type="text"
                    id="recipeName"
                    name="name"
                    className="form-control"
                    placeholder="The name of the dish..."
                    value={newRecipe.name}
                    onChange={this.onChange}
                  />
                  {
                    errors.name &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.name}
                    </span>
                  }
                </div>


                <div className="form-group form-width">
                  <input
                    type="text"
                    id="recipeIngredients"
                    name="ingredients"
                    className="form-control"
                    placeholder="A comma seperated list of ingredients..."
                    value={newRecipe.ingredients}
                    onChange={this.onChange}
                  />
                  {
                    errors.ingredients &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.ingredients}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <textarea
                    type="text"
                    rows="5"
                    id="recipeDescription"
                    className="form-control"
                    placeholder="Instructions on how this dish is made..."
                    name="description"
                    value={newRecipe.description}
                    onChange={this.onChange}
                  />
                  {
                    errors.description &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.description}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <button
                    id="recipeButton"
                    className="btn btn-secondary"
                    disabled={this.state.submitting}
                  >
                    {this.state.submitting ? 'Submitting...' : 'Add Recipe'}
                  </button>
                  <Link
                    className="btn btn-secondary"
                    to="/users/recipes"
                  >
                    Cancel
                  </Link>
                </div>
              </form>
            </div>
          </div>
        </section>
      </div>

    );
  }
}

AddRecipe.propTypes = {
  addNewRecipe: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};


const mapStateToProps = state => ({
  addRecipe: state.addRecipe,
  authenticated: state.auth.isAuthenticated,
  errorMessage: state.errorMessage
});

const ConnectedComponent =
  connect(mapStateToProps, { addNewRecipe: addRecipe })(AddRecipe);
export { ConnectedComponent as ConnectedAddRecipe };

