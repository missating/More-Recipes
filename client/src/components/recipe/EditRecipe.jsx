import React from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import { Link, Redirect } from 'react-router-dom';

// validations
import recipeValidator from '../../validation/recipeValidator';

// action
import getSingleRecipe from '../../actions/getSingleRecipe';
import editRecipe from '../../actions/editRecipe';


/**
 * @description Creates EditRecipe component
 *
 * @class EditRecipe
 *
 * @extends {React.Component}
 */
export class EditRecipe extends React.Component {
  /**
   * @description Creates an instance of EditRecipe.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      ingredients: '',
      description: '',
      recipeImage: '',
      newImage: '',
      updating: false,
      errors: {},
    };
    this.onChange = this.onChange.bind(this);
    this.onEdit = this.onEdit.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
  }


  /**
   * @description Get's a recipe
   *
   * @method
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  componentDidMount() {
    const recipeId = this.props.match.params.id;
    this.props.getRecipeDetails(recipeId);
  }


  /**
   * @description Sets state
   *
   * @param {object} nextProps
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    const {
      id, name, description, ingredients, recipeImage
    } = nextProps.singleRecipe;
    this.setState({
      id, name, description, ingredients, recipeImage
    });
  }


  /**
   * @description Bind the value of the inputs to state
   *
   * @method onChange
   *
   * @param {object} event
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }


  /**
   * @description Update's the Recipe
   *
   * @method onEdit
   *
   * @param {object} event
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  onEdit(event) {
    event.preventDefault();
    this.setState({
      updating: true
    });
    const {
      id, name, description, ingredients, newImage, recipeImage
    } = this.state;
    const isValid = this.isValid();
    if (isValid) {
      if (newImage) {
        this.uploadToCloudinary().then((response) => {
          const secureUrl = response.data.secure_url;
          this.props.updateRecipe(
            {
              name, description, ingredients, recipeImage: secureUrl
            },
            id
          ).then(() => {
            this.props.history.push('/users/recipes');
          });
        });
      } else {
        this.props.updateRecipe(
          {
            name, description, ingredients, recipeImage
          },
          id
        ).then(() => {
          this.props.history.push('/users/recipes');
        });
      }
    }
  }

  /**
   * @description Handles image drop
   *
   * @method handleDrop
   *
   * @memberof EditRecipe
   *
   * @param {array} files
   *
   * @returns {void}
   */
  handleDrop(files) {
    this.setState({
      newImage: files[0],
      recipeImage: files[0].preview
    });
  }

  /**
   * @description Handle image upload
   *
   * @method uploadToCloudinary
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  uploadToCloudinary() {
    const formData = new FormData();
    formData.append('file', this.state.newImage);
    formData.append('upload_preset', 'moreRecipes');
    formData.append('api_key', '462882972372719');

    return axios.post(
      'https://api.cloudinary.com/v1_1/dxayftnxb/image/upload',
      formData, {
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
      }
    );
  }

  /**
    * @description Validates user's data before making post request
    *
    * @method isValid
    *
    * @memberof EditRecipe
    *
    * @returns {boolean} true or false
    */
  isValid() {
    const { errors, isValid } = recipeValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    } else {
      this.setState({ errors: {} });
      return isValid;
    }
  }

  /**
    * @description Renders react component
    *
    * @method render
    *
    * @memberof EditRecipe
    *
    * @returns {void}
    *
    */
  render() {
    return (
      <div>
        {
          !this.props.authenticated &&
          <Redirect to="/" />
        }

        <section className="container" id="recipes" >

          <h3 className="text-center mb-4">Edit Recipe</h3>

          <div className="row">
            <div className="col-md-12 col-sm-12">
              <form>
                <Dropzone
                  onDrop={this.handleDrop}
                  accept="image/*"
                  multiple={false}
                  style={{
                    border: '3px dashed ',
                    width: '300px',
                    height: 'auto',
                    margin: '0 auto',
                    padding: '5px'
                  }}
                >
                  {
                    this.state.recipeImage &&
                    <div>
                      <img
                        src={this.state.recipeImage}
                        alt=""
                        className="img-fluid mx-auto d-block"
                      />
                      <h5 className="text-center">
                        Click here to upload
                      </h5>
                    </div>
                  }
                </Dropzone>

                <div className="form-group form-width">
                  <h5> Name </h5>
                  <input
                    type="text"
                    className="form-control"
                    name="name"
                    value={this.state.name}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.name &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.name}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <h5> Ingredients </h5>
                  <input
                    type="text"
                    className="form-control"
                    name="ingredients"
                    value={this.state.ingredients}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.ingredients &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.ingredients}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <h5> Description </h5>
                  <textarea
                    type="text"
                    rows="5"
                    className="form-control"
                    id="updateDescription"
                    name="description"
                    value={this.state.description}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.description &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.description}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <button
                    id="editButton"
                    className="btn btn-secondary"
                    disabled={this.state.updating}
                    onClick={this.onEdit}
                  >
                    {this.state.updating ? 'Updating...' : 'Update'}
                  </button>

                  <Link
                    className="btn btn-secondary"
                    to="/users/recipes"
                  >
                    Cancel
                  </Link>
                </div>
              </form>
            </div>
          </div>
        </section >
      </div >
    );
  }
}


EditRecipe.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  singleRecipe: PropTypes.shape({}).isRequired,
  getRecipeDetails: PropTypes.func.isRequired,
  updateRecipe: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};


const mapStateToProps = state => ({
  singleRecipe: state.recipes.singleRecipe,
  authenticated: state.auth.isAuthenticated
});

const mapDispatchToProps = dispatch => ({
  getRecipeDetails: id => dispatch(getSingleRecipe(id)),
  updateRecipe: (recipe, id) => dispatch(editRecipe(recipe, id))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(EditRecipe);
export { ConnectedComponent as ConnectedEditRecipe };

