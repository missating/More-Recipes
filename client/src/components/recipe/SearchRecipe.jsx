import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import searchRecipes from '../../actions/searchRecipe';


/**
 * @description Creates SearchRecipe component
 *
 * @class SearchRecipe
 *
 * @extends {React.Component}
 */
export class SearchRecipe extends React.Component {
  /**
  * @description Creates an instance of SearchRecipe.
  *
  * @constructor
  *
  * @param {object} props
  *
  * @memberof SearchRecipe
  *
  * @returns {void}
  */
  constructor(props) {
    super(props);
    this.state = {
      searchQuery: ''
    };
    this.onChange = this.onChange.bind(this);
  }

  /**
  * @description Bind the value of the inputs to state
  *
  * @method onChange
  *
  * @param {object} event
  *
  * @memberof SearchRecipe
  *
  * @returns {void}
  */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
    this.props.searchRecipes(
      event.target.value,
      this.props.pagination.currentPage
    );
  }

  /**
    * @description Renders react component
    *
    * @method render
    *
    * @memberof SearchRecipe
    *
    * @returns {void}
    *
    */
  render() {
    return (
      <div>
        <div className="col-lg-12">
          <form>
            <div className="input-group">
              <input
                id="search"
                type="text"
                name="searchQuery"
                className="form-control"
                value={this.state.searchQuery}
                onChange={this.onChange}
                placeholder="Search for..."
              />
            </div>
          </form>
        </div >
        <hr />
      </div >

    );
  }
}

SearchRecipe.propTypes = {
  searchRecipes: PropTypes.func.isRequired,
  pagination: PropTypes.shape({
    currentPage: PropTypes.number
  })
};

SearchRecipe.defaultProps = {
  pagination: {
    currentPage: 1
  }
};

const mapDispatchToProps = dispatch => ({
  searchRecipes: searchQuery => (dispatch(searchRecipes(searchQuery)))
});

const mapStateToProps = state => ({
  pagination: state.pagination,
});


const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(SearchRecipe);
export { ConnectedComponent as ConnectedSearchRecipe };
