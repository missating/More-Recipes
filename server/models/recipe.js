export default (sequelize, DataTypes) => {
  const Recipe = sequelize.define('Recipe', {
    name: {
      type: DataTypes.STRING,
      allowNull: false
    },

    userId: {
      type: DataTypes.INTEGER,
      allowNull: false
    },

    ingredients: {
      type: DataTypes.STRING,
      allowNull: false
    },

    description: {
      type: DataTypes.TEXT,
      allowNull: false
    },

    recipeImage: {
      type: DataTypes.STRING,
      allowNull: true
    },


    favourite: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0
    },

  });

  Recipe.associate = (models) => {
    // associations defined here
    Recipe.belongsTo(
      models.User,
      { foreignKey: 'userId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Review,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Vote,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Favourite,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
  };
  return Recipe;
};
