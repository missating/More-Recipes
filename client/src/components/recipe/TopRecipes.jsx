import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// components
import RecipeCard from '../cards/RecipeCard';

// actions
import getTopRecipes from '../../actions/getTopRecipes';


/**
 * @description Creates TopRecipes component
 *
 * @class TopRecipes
 *
 * @extends {React.Component}
 */
export class TopRecipes extends React.Component {
  /**
   * @description Gets top recipes
   *
   * @method
   *
   * @memberof TopRecipes
   *
   * @returns {void}
   */
  componentWillMount() {
    this.props.recipes();
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof TopRecipes
   *
   * @returns {void}
   */
  render() {
    const topRecipe = (this.props.topRecipes) ? this.props.topRecipes : [];

    let topRecipeList = topRecipe.map((recipe, i) => (
      <div className="col-md-4" key={`topRecipe${i + 1}`}>
        <RecipeCard
          {...recipe}
        />

      </div>
    ));

    if (topRecipe.length === 0) {
      topRecipeList =
        (
          <div>
            <h4 className="text-center display-5">
              Currently no featured recipes
            </h4>
          </div>
        );
    }
    return (
      <div>
        <section className="container" id="featured">
          <h2 className="text-center mt-5"> Featured Recipes </h2>
          <hr />
          <div className="row justify-content-center">
            {topRecipeList}
          </div>
        </section>
      </div>
    );
  }
}

TopRecipes.propTypes = {
  recipes: PropTypes.func.isRequired,
  topRecipes: PropTypes.arrayOf(PropTypes.object)
};

TopRecipes.defaultProps = {
  topRecipes: []
};


const mapStateToProps = state => ({
  topRecipes: state.topRecipes.recipes
});

const mapDispatchToProps = dispatch => ({
  recipes: () => dispatch(getTopRecipes())
});


const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(TopRecipes);
export { ConnectedComponent as ConnectedTopRecipes };

