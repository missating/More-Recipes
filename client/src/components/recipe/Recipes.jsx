import React from 'react';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import PropTypes from 'prop-types';

// components
import RecipeCard from '../cards/RecipeCard';
import { ConnectedSearchRecipe } from '../recipe/SearchRecipe';

// actions
import getAllRecipes from '../../actions/getAllRecipes';

/**
 * @description Creates Recipes component
 *
 * @class Recipes
 *
 * @extends {React.Component}
 */
export class Recipes extends React.Component {
  /**
   * @description Creates an instance of Recipes.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
    };
    this.onPageChange = this.onPageChange.bind(this);
  }


  /**
   * @description Get recipes
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  componentWillMount() {
    this.props.recipes();
  }

  /**
   * @description pagination for recipes
   *
   * @param {number} current
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  onPageChange(current) {
    const selected = current.selected + 1;
    this.props.history.push(`/recipes?page=${selected}`);
    this.props.recipes(selected);
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof Recipes
   *
   * @returns {void}
   *
   */
  render() {
    const { pages } = this.props.pagination;

    const recipes = this.props.allRecipes ? this.props.allRecipes : [];
    let recipeList;
    if (recipes.length === 0) {
      recipeList = (
        <div className="text-center ml-5" style={{ width: '100%' }}>
          <h4> No recipe(s) found </h4>
        </div>
      );
    } else {
      recipeList = recipes.map((recipe, i) => (
        <div className="col-md-6 col-lg-4 p-0" key={`recipe${i + 1}`}>
          <RecipeCard
            {...recipe}
          />
        </div>
      ));
    }


    return (
      <div>
        <section className="container" id="recipes">

          <ConnectedSearchRecipe />

          <div className="row mh-500">
            {recipeList}
          </div>


          <div className="container">
            <ReactPaginate
              pageCount={pages}
              pageRangeDisplayed={5}
              marginPagesDisplayed={3}
              previousLabel="Previous"
              nextLabel="Next"
              breakClassName="text-center"
              initialPage={0}
              containerClassName="container pagination justify-content-center"
              pageClassName="page-item"
              pageLinkClassName="page-link"
              activeClassName="page-item active"
              previousClassName="page-item"
              nextClassName="page-item"
              nextLinkClassName="page-link"
              previousLinkClassName="page-link"
              onPageChange={this.onPageChange}
            />
          </div>


        </section>
      </div>
    );
  }
}

Recipes.propTypes = {
  recipes: PropTypes.func.isRequired,
  allRecipes: PropTypes.arrayOf(PropTypes.object),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  pagination: PropTypes.shape({
    pages: PropTypes.number
  })
};

Recipes.defaultProps = {
  pagination: {
    pages: 1
  },
  allRecipes: []
};

const mapStateToProps = state => ({
  allRecipes: state.recipes.allrecipes,
  pagination: state.pagination,
  isFetching: state.isFetching,
});

const mapDispatchToProps = dispatch => ({
  recipes: page => dispatch(getAllRecipes(page))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(Recipes);
export { ConnectedComponent as ConnectedRecipes };

