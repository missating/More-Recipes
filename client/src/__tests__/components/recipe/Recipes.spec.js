import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import { Recipes } from '../../../components/recipe/Recipes';
import RecipeCard from '../../../components/cards/RecipeCard';

describe('Recipe component', () => {
  const props = {
    history: {
      push: jest.fn()
    }
  };
  const getAllRecipes = jest.fn(() => Promise.resolve());

  it('Should render', () => {
    const wrapper = shallow(<Recipes
      recipes={getAllRecipes}
      allRecipes={[]}
      pagination={{ pages: 1 }}
      {...props}
    />);
    expect(wrapper.find(RecipeCard).exists).toBeTruthy();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('Should not render recipes if allRecipes is an empty array', () => {
    const wrapper = shallow(<Recipes
      recipes={getAllRecipes}
      allRecipes={[]}
      pagination={{ pages: 1 }}
      {...props}
    />);
    expect(wrapper.find(RecipeCard).length).toEqual(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('Should render 2 recipes if allRecipes contains 2 objects', () => {
    const allRecipes = [
      {
        id: 1,
        recipeImage: 'image_url',
        name: 'Eba',
        ingredients: 'Garri, Fire',
        description: 'Hello, this is my favorite food'
      },
      {
        id: 2,
        recipeImage: 'image_url',
        name: 'Eba',
        ingredients: 'Garri, Fire',
        description: 'Hello, this is my favorite food'
      }
    ];
    const wrapper = shallow(<Recipes
      recipes={getAllRecipes}
      allRecipes={allRecipes}
      pagination={{ pages: 1 }}
      {...props}
    />);
    expect(wrapper.find(RecipeCard).length).toEqual(2);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should set on page change ', () => {
    const spy = sinon.spy();

    const selected = {
      current: 1
    };
    const wrapper = shallow(<Recipes
      recipes={spy}
      allRecipes={[]}
      pagination={{ pages: 3 }}
      {...props}
    />);
    wrapper.instance().onPageChange(selected);
    expect(spy.called).toBeTruthy();
  });
});
