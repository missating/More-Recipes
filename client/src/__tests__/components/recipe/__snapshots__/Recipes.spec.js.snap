// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe component Should not render recipes if allRecipes is an empty array 1`] = `
<div>
  <section
    className="container"
    id="recipes"
  >
    <Connect(SearchRecipe) />
    <div
      className="row mh-500"
    >
      <div
        className="text-center ml-5"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <h4>
           No recipe(s) found 
        </h4>
      </div>
    </div>
    <div
      className="container"
    >
      <PaginationBoxView
        activeClassName="page-item active"
        breakClassName="text-center"
        breakLabel="..."
        containerClassName="container pagination justify-content-center"
        disableInitialCallback={false}
        disabledClassName="disabled"
        initialPage={0}
        marginPagesDisplayed={3}
        nextClassName="page-item"
        nextLabel="Next"
        nextLinkClassName="page-link"
        onPageChange={[Function]}
        pageClassName="page-item"
        pageCount={1}
        pageLinkClassName="page-link"
        pageRangeDisplayed={5}
        previousClassName="page-item"
        previousLabel="Previous"
        previousLinkClassName="page-link"
      />
    </div>
  </section>
</div>
`;

exports[`Recipe component Should render 1`] = `
<div>
  <section
    className="container"
    id="recipes"
  >
    <Connect(SearchRecipe) />
    <div
      className="row mh-500"
    >
      <div
        className="text-center ml-5"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <h4>
           No recipe(s) found 
        </h4>
      </div>
    </div>
    <div
      className="container"
    >
      <PaginationBoxView
        activeClassName="page-item active"
        breakClassName="text-center"
        breakLabel="..."
        containerClassName="container pagination justify-content-center"
        disableInitialCallback={false}
        disabledClassName="disabled"
        initialPage={0}
        marginPagesDisplayed={3}
        nextClassName="page-item"
        nextLabel="Next"
        nextLinkClassName="page-link"
        onPageChange={[Function]}
        pageClassName="page-item"
        pageCount={1}
        pageLinkClassName="page-link"
        pageRangeDisplayed={5}
        previousClassName="page-item"
        previousLabel="Previous"
        previousLinkClassName="page-link"
      />
    </div>
  </section>
</div>
`;

exports[`Recipe component Should render 2 recipes if allRecipes contains 2 objects 1`] = `
<div>
  <section
    className="container"
    id="recipes"
  >
    <Connect(SearchRecipe) />
    <div
      className="row mh-500"
    >
      <div
        className="col-md-6 col-lg-4 p-0"
        key="recipe1"
      >
        <RecipeCard
          button={false}
          description="Hello, this is my favorite food"
          id={1}
          ingredients="Garri, Fire"
          name="Eba"
          onButtonClick={[Function]}
          recipeImage="image_url"
        />
      </div>
      <div
        className="col-md-6 col-lg-4 p-0"
        key="recipe2"
      >
        <RecipeCard
          button={false}
          description="Hello, this is my favorite food"
          id={2}
          ingredients="Garri, Fire"
          name="Eba"
          onButtonClick={[Function]}
          recipeImage="image_url"
        />
      </div>
    </div>
    <div
      className="container"
    >
      <PaginationBoxView
        activeClassName="page-item active"
        breakClassName="text-center"
        breakLabel="..."
        containerClassName="container pagination justify-content-center"
        disableInitialCallback={false}
        disabledClassName="disabled"
        initialPage={0}
        marginPagesDisplayed={3}
        nextClassName="page-item"
        nextLabel="Next"
        nextLinkClassName="page-link"
        onPageChange={[Function]}
        pageClassName="page-item"
        pageCount={1}
        pageLinkClassName="page-link"
        pageRangeDisplayed={5}
        previousClassName="page-item"
        previousLabel="Previous"
        previousLinkClassName="page-link"
      />
    </div>
  </section>
</div>
`;
